Documentation guide: Solve one or a system of inequalities

bertiewooster opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed


#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* other
  * A new, user-friendly page on solving a system of inequalities algebraically was added to the Guides section.
<!-- END RELEASE NOTES -->

bertiewooster committed:
Start page

bertiewooster committed:
Add some examples

bertiewooster committed:
Add examples

bertiewooster committed:
Add examples

bertiewooster committed:
Add examples

bertiewooster committed:
Link from solving index to new sub-page

bertiewooster commented:
This is currently a very rough draft. Anything in *italics* is to be filled in, for example links to other pages in the documentation.

At this point, I'd appreciate content suggestions, for example are there any speed vs. reliability tradeoffs to include.

bertiewooster committed:
Fix two doctests

bertiewooster committed:
Sort relations to ensure consistent list order for
docstring testing

bertiewooster committed:
Add new sub-page to (hidden) toc
on main page so Sphinx can sequence sub-page

bertiewooster committed:
Fix exception doctests by
putting entire error message on one line

bertiewooster committed:
Fill in links

bertiewooster committed:
Link to square roots as
What is symbolic computation?

bertiewooster commented on doc/src/guides/solving/solve-system-of-inequalities-algebraically.md:
```diff
@@ -0,0 +1,229 @@
+We recommend you include the variable to be solved for as the second argument 
+for {func}`~.solve`. While {func}`~.solve` can currently solve systems of equations with only one symbol, it is a good practice in case that capability is expanded, and because {func}`~.solve` can solve an equation with more than one symbol.
+
+### Use exact values
```
I'm wondering whether this section should be in a general solving guidance page--this (and probably other guidance) could apply to almost any type of solving.

oscarbenjamin replied:
Yes, that's probably better. Also parsing strings probably doesn't need to be repeated in lots of places and could go in a separate guide.

bertiewooster replied:
Moved to #23777

bertiewooster commented on doc/src/guides/solving/solve-system-of-inequalities-algebraically.md:
````diff
@@ -0,0 +1,229 @@
+[1/3, sqrt(pi)]
+```
+
+## *Tradeoffs (speed vs. accuracy, etc.) for function*
````
Are `check=False` and `simplify=False` worth mentioning here as ways to potentially speed up `solve` for a system of inequalities?

I am also wondering whether it makes sense to have a page that describes such options for `solve` in general, rather than discussing those options on multiple solving pages within doc/guides/solving.

oscarbenjamin replied:
I don't know if check and simplify have any effect for inequalities.

bertiewooster replied:
Thanks.

oscarbenjamin commented on doc/src/guides/solving/solve-system-of-inequalities-algebraically.md:
```diff
@@ -0,0 +1,229 @@
+If you want to preserve the exact mathematical values of symbols such as
+[fractions](tutorial-gotchas-final-notes) and [square
+roots](symbolic-computation), define them so that SymPy can interpret them
+symbolically, for example define one-third as `Integer(1)/Integer(3)`:
```
Better suggestions are `S(1)/3` or `Rational(1, 3)`.

oscarbenjamin commented on doc/src/guides/solving/solve-system-of-inequalities-algebraically.md:
```diff
@@ -0,0 +1,229 @@
+# Solve a system of inequalities algebraically
+
+Use SymPy to solve a system of inequalities algebraically. For example, solving
+$x^2 < \pi$, $x > 0$ yields $0 < x < \sqrt{pi}$.
```
I would expect a system of inequalities to mean a *multivariate* system of inequalities involving multiple unknowns. Using `solve` like this only works for *univariate* inequalities e.g.:
```
In [2]: solve([x > y**2, y > 0], [x, y])
---------------------------------------------------------------------------
NotImplementedError
```
The fact that this does not handle multivariate inequalities should be made clear in the title

bertiewooster commented on doc/src/guides/solving/solve-system-of-inequalities-algebraically.md:
```diff
@@ -0,0 +1,229 @@
+
+Alternatives to consider:
+- For systems with more than one symbol (variable), try [Wolfram
+Alpha](https://www.wolframalpha.com/)
```
I debated whether to include such links. Is there some open-source way to solve multivariate systems of inequalities? (I coulnd't find one; it seems like SymPy came close in #21687 in Fall 2021, yet the pull request wasn't comlpeted.) If not, is it better to provide links to a closed-source, free option that will help visitors solve their problem, or not provide a solution?

oscarbenjamin replied:
There is:
https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.linprog.html

Another SymPy PR: #22389

SymPy can definitely be improved in this respect. I would rather link to an issue describing how to improve SymPy than to Wolfram Alpha though.

bertiewooster replied:
Thanks

bertiewooster committed:
Address several review comments

bertiewooster committed:
Remove "Use exact values" section
Will put in Solving Guidance instead

bertiewooster committed:
Set guides/index toc maxdepth to 1

bertiewooster committed:
Replace all references to Wolfram Alpha
with SciPy linear  programming function

bertiewooster committed:
Remove section on parsing--
in solving guidance instead

bertiewooster committed:
Merge branch 'master' into system-of-inequalities
To incorporate changes from first solving sub-page.

bertiewooster committed:
Merge branch 'master' into system-of-inequalities

asmeurer commented on doc/src/guides/solving/index.md:
```diff
@@ -16,7 +16,7 @@ Learn how to use SymPy computer algebra system to:
 |  {func}`Solve a matrix problem algebraically <sympy.matrices.matrices.MatrixBase.solve>`                    | $ \left[\begin{array}{cc} 1 & 1\\1 & -1\end{array}\right] \left[\begin{array}{cc} x\\y\end{array}\right] = \left[\begin{array}{cc} 2\\0\end{array}\right] $ | $ \left[\begin{array}{cc} x\\y\end{array}\right] = \left[\begin{array}{cc} 1\\1\end{array}\right]$  |
 |  [ Solve an inequality algebraically ](../../modules/solvers/inequalities.rst)                      | $ x^2 < 4 $ | $ -2 < x < 2 $                                                                                                        |
-|  [ Solve a system of inequalities algebraically ](../../modules/solvers/inequalities.rst)           | $ x^2 < 4, x > 0 $ | $ 0 < x < 2 $                                                                                                  |
+|  [ Solve a system of inequalities algebraically ](solve-system-of-inequalities-algebraically.md)           | $ x^2 < \pi, x > 0 $ | $ 0 < x < \sqrt{pi} $                                                                                                  |
```
There's a missing backslashes on `pi` in many equations on this page.

asmeurer commented on doc/src/guides/solving/solve-system-of-inequalities-algebraically.md:
```diff
@@ -0,0 +1,186 @@
+(variable). For example, for a solution of $0 < x < \sqrt{pi}$, you might want
+to extract $0$ and $\sqrt{pi}$.
+
+### Extract relational atoms
```
I think we should just omit this section. No one should really use Relational directly like this.

oscarbenjamin commented:
SymPy does not really support solving a system of inequalities in any meaningful way. If we're going to have a docs page about how to do this then it needs to be up front about that fact that SymPy's capability is extremely limited. If I saw a page called "solve a system of inequalities" then I would expect it to show examples of doing something with a multivariate system of inequalities but SymPY basically can't do anything with them:
```
In [10]: solve([x + y > 1, y > 0], [x, y])
---------------------------------------------------------------------------
NotImplementedError
```
Also every example in this page shows using solve in a way that I wish we could just deprecate or change. Ideally solve with inequalities would not behave in the way that is shown in these examples.

bertiewooster committed:
Remove suggestion to create relations--
no reason for someone to

bertiewooster committed:
Use less technical titles for Extract sections

bertiewooster commented:
> SymPy does not really support solving a system of inequalities in any meaningful way. If we're going to have a docs page about how to do this then it needs to be up front about that fact that SymPy's capability is extremely limited.

Concur; I changed the page title to "Solve a system of inequalities _of a single variable_ algebraically".

In addition to SymPy's inability to handle multivariate systems of inequalities, what other limitations should we state up front? cc @smichr

bertiewooster committed:
Add note box at top:
SymPy limited to inequalities of one variable

smichr commented:
Another limitation is that `solve` does not work out Boolean expressions, but the `as_set` method can do so:
```python
>>> And(And(x>1,x<3),And(x>2,x<4)).as_set().as_relational(x)
(2 < x) & (x < 3)
```

smichr commented on doc/src/guides/solving/solve-system-of-inequalities-algebraically.md:
````diff
@@ -0,0 +1,167 @@
+```py
+>>> from sympy import symbols, solve, pi, Integer
+>>> x = symbols('x')
+>>> one_third = Integer(1)/Integer(3)
+>>> solve([x >= one_third, x**2 <= pi], x)
````
```suggestion
>>> solve([3*x >= 1, x**2 <= pi], x)
```

smichr commented on doc/src/guides/solving/solve-system-of-inequalities-algebraically.md:
```diff
@@ -0,0 +1,167 @@
+>>> from sympy import symbols, solve, Integer, pi
+>>> from sympy.core.relational import Relational
+>>> x = symbols('x')
+>>> one_third = Integer(1)/Integer(3)
```
same here; otherwise consider using the more ubiquitous `S(1)/3` or `Rational(1, 3)`.

smichr commented on doc/src/guides/solving/solve-system-of-inequalities-algebraically.md:
````diff
@@ -0,0 +1,167 @@
+```py
+>>> from sympy import symbols, solve, Integer, pi
+>>> x = symbols('x')
+>>> eq = solve([x >= Integer(1)/Integer(3), x**2 <= pi], x)
````
ditto

smichr commented on doc/src/guides/solving/solve-system-of-inequalities-algebraically.md:
```diff
@@ -0,0 +1,167 @@
+### Equations which have an analytical solution, and SymPy cannot solve
+
+SymPy has implemented algorithms to solve inequalities involving only one symbol
+(variable), so it cannot solve a set of inequalities involving more than one
```
It might be better to indicate that it can solve for a single symbol; this might be removed from `solve` but will always be available through `reduce_inequalities`:
```python
>>> solve([x + y > 1, y > 0], x)
(0 < y) & (y < oo) & (x > 1 - y)
>>> reduce_inequalities([x + y > 1, y > 0], x)
(0 < y) & (y < oo) & (x > 1 - y)
```
But the final relationship, `x < 1` is not currently resolved.

bertiewooster replied:
Thanks. I'm having trouble understanding the difference between solving for just `x` in the above and solving for `x` and `y`--what would be the result of solving for x and y?

Also, do solve and reduce_inequalities work for only linear inequalities? I ask because I get this error if I include a symbol to the second power (or third or fourth):

```
>>>solve([x ** 2 < y, x > 0], x)
NotImplementedError: 
The inequality, -_y + x**2 < 0, cannot be solved using
solve_univariate_inequality.
```

asmeurer replied:
`(0 < y) & (y < oo) & (x > 1 - y)` implies that `y` is in `(0, oo)`, so `x > 1 - y` would mean `x` is in `(-oo, 1)`, so a more complete solution would be `(x < 1) & (y > 0)`.

Also setting `finite=True` would remove the infinity inequalities, which might be worth mentioning in this guide.

bertiewooster replied:
Thanks. So SymPy can currently reduce a system of inequalities _containing_ more than one symbol, but cannot _reduce for_ more than one symbol. (The phrase "solve for" seems more natural, but we're trying to avoid calling this solving. Although the functions themselves are within `sympy.solvers` and return errors mentioning e.g. "solved"...)

bertiewooster replied:
I'm not including `finite=True` because it seems to be inapplicable for `reduce_inequalities`.

bertiewooster committed:
Use intersphinx links to SciPy
Link to reduce Boolean expressions

bertiewooster committed:
Replace x > one-third with 3*x > 1

bertiewooster commented:
> I tend to think that using `solve` in the way this guide suggests should probably be deprecated in favour of the explicit inequality solve methods.

@oscarbenjamin so do you think that this guide should use `reduce_inequalities` instead of `solve`?

And is `reduce_inequalities` the top-level reduce function that will call [other inequality solvers](https://docs.sympy.org/dev/modules/solvers/inequalities.html) (`solve_rational_inequalities`, `solve_poly_inequality`, etc.) as needed, in the way that `solve` is the top-level solving function that will call lower-level solving functions as needed? When I was new to SymPy and trying to solve a system of inequalities, I was very confused about which of those inequality solvers I should use. When I discovered that `solve` would work, I went with that.

bertiewooster committed:
Merge branch 'master' into system-of-inequalities

smichr commented:
> And is reduce_inequalities the top-level reduce function

yes. And `solve`, as soon as it detects and Relational calls that function.
https://github.com/sympy/sympy/blob/353653b7a3afc4ce2f25c1698efe4544ba9312f8/sympy/solvers/solvers.py#L957-L959

oscarbenjamin commented:
I think it would be better to recommend `reduce_inequalities`. The name also better clarifies what the function does. I don't think it makes sense to say "solve a system of inequalities" or at least that's not how I would describe what `solve` or `reduce_inequalities` actually do.

bertiewooster committed:
Change solve() to reduce_inequalities()
Add intersphix link to scipy

bertiewooster committed:
Clarify one-variable limit is reducing for,
not system-of-inequalities containing

bertiewooster committed:
Tweak wording

bertiewooster committed:
Added other limitations

bertiewooster committed:
Fix link to a module page
Try to fix 'could not lex literal_block as "py"'

bertiewooster committed:
Removed py from ``` code block
to try to prevent "could not lex as py" error

bertiewooster commented:
For some reason, this error was thrown by CircleCI:

`/home/circleci/repo/doc/src/guides/solving/solve-system-of-inequalities-algebraically.md:87: WARNING: Could not lex literal_block as "py". Highlighting skipped.`

for this docstring

````
```py
>>> from sympy import reduce_inequalities
>>> from sympy.abc import x, y
>>> reduce_inequalities([x ** 2 < 4, x > 0], x)
(0 < x) & (x < 2)
>>> reduce_inequalities([x < y, x > 0], x)
(0 < x) & (x < oo) & (x < y)
>>> reduce_inequalities([x ** 2 - y < 4, x > 0], x)
Traceback (most recent call last):
    ...
NotImplementedError: 
The inequality, -_y + x**2 - 4 < 0, cannot be solved using
solve_univariate_inequality.
```
````

I removed the `py` from ```` ```py```` and CircleCI was OK with it. I'm not sure what the problem is because it built fine on my computer, and the code runs fine in a Jupyter Notebook.

If this is a problem, I'd appreciate any suggestions.

bertiewooster committed:
Shorten function display name to as_set

bertiewooster committed:
Use Title Case for headings

bertiewooster committed:
Remove "simple"

oscarbenjamin commented on doc/src/guides/solving/solve-system-of-inequalities-algebraically.md:
````diff
@@ -0,0 +1,218 @@
+```py
+>>> from sympy import reduce_inequalities, symbols
+>>> x, y = symbols("x y")
+>>> reduce_inequalities([x + y > 1, y > 0], x)
+(0 < y) & (y < oo) & (x > 1 - y)
+```
+
+However, SymPy cannot arrive at the more complete result `(x < 1) & (y > 0)`.
````
The solution given by `reduce_inequalities` is complete.

bertiewooster replied:
I was using the language from @asmeurer

> (0 < y) & (y < oo) & (x > 1 - y) implies that y is in (0, oo), so x > 1 - y would mean x is in (-oo, 1), so a more complete solution would be (x < 1) & (y > 0).

Is there a better way to express this limitation (if you agree it is one) of `reduce_inequalities`?

oscarbenjamin replied:
The result `(x < 1) & (y > 0)` is not implied by the original inequalities. As mentioned by @rdorsch above `x = 5` and `y = 5` would satisfy the original inequalities but clearly would not satisfy `x < 1`.

bertiewooster replied:
Removed that note about "complete result."

bertiewooster committed:
Changes per comments

bertiewooster committed:
Clarify that SymPy can reduce a system for multiple symbols, but only if there's only one symbol per inequality

bertiewooster committed:
Put all limitations content in one section, link to that from Inequalities Which Can Be Reduced Analytically, and SymPy Cannot Reduce

bertiewooster committed:
Add/correct Sphinx links

bertiewooster commented:
If there are no other comments on this pull request (the last comment was two weeks ago), it's ready to be merged. The new page it adds is [Reduce a System of Inequalities for a Single Variable Algebraically](https://output.circle-artifacts.com/output/job/d96e90a4-6c45-417f-a0a8-36669aff3f8c/artifacts/0/doc/_build/html/guides/solving/solve-system-of-inequalities-algebraically.html).

asmeurer commented:
> /home/circleci/repo/doc/src/guides/solving/solve-system-of-inequalities-algebraically.md:87: WARNING: Could not lex literal_block as "py". Highlighting skipped.

I'm not sure but the error could be because of the indented ellipsis `...`. It typically should not be indented.

bertiewooster committed:
Remove indent before ellipsis in code blocks
Replace print(<expr>) with <expr>

bertiewooster committed:
Explain args is arguments, oo is infinity

bertiewooster committed:
Remove py from code block to prevent error
Could not lex literal_block as "py"

bertiewooster commented:
> I'm not sure but the error could be because of the indented ellipsis `...`. It typically should not be indented.

Thanks for the suggestion. I removed the indentation from the ellipsis and the problem persisted. I removed the `py` from ```` ```py```` again and it prevented the error. Updated new page is [Reduce a System of Inequalities for a Single Variable Algebraically](https://output.circle-artifacts.com/output/job/19d0ab21-eace-46e5-84df-476d55cdb502/artifacts/0/doc/_build/html/guides/solving/solve-system-of-inequalities-algebraically.html).

asmeurer commented:
I think it's not highlighting because of whatever the error is. Usually it guesses the programming language automatically so you don't actually have to include the `py`.

bertiewooster committed:
Remove extraneous line break in code block error trace

bertiewooster commented:
Turns out the problem was a line break in the traceback, after the error type `NotImplementedError:` After removing that line break to give

```
NotImplementedError: The inequality, -_y + x**2 - 4 < 0, cannot be solved using
solve_univariate_inequality.
```

 and adding `py` back into ```` ```py````,   version builds without error and code is highlighted correctly. So that problem is resolved.

For future reference, I learned that you can tell on your local build of the docs whether there will be a problem on CircleCI build: if highlighting is not present on your local build of the docs, will probably throw an error on CircleCI build. So I was able to check that my changes fixed the problem locally, then I pushed the commit to GitHub. Checking locally is advantageous because it takes a while (about eight minutes, I believe) to build the docs on CircleCI after you push the commit.

asmeurer commented:
Your local build should also have all the same errors as would appear on CircleCI, if you look at the build log. Sometimes errors only appear on the first build and won't show up again unless you do a `make clean` and rebuild.

bertiewooster commented:
If there are no final comments on this page, it's ready to be merged.

bertiewooster committed:
Merge branch 'master' into system-of-inequalities

bertiewooster committed:
Close toctree directive on main solving page

bertiewooster committed:
Retitle sub-page, make Alternatives to Consider a heading

bertiewooster committed:
Include one inequality as well as a system

asmeurer commented:
This looks good. Just want to double check that it can be merged since it is in draft status.

bertiewooster commented:
@asmeurer thanks for reviewing. I just wanted to add a note linking to the existing [inequalities module](https://docs.sympy.org/latest/modules/solvers/inequalities.html) linking to this page, which should just take a few minutes when I have time. Or if you'd like to merge this as is, that's fine too, I could make a separate pull request for the note.

bertiewooster committed:
Add link from inequalities solver module to guide

bertiewooster commented:
@asmeurer I just pushed that change.

bertiewooster committed:
Merge branch 'master' into system-of-inequalities
